### DISCUSSION TITLE
How does the input, when-clause context, and command system work?

### DISCUSSION THREAD

**Discussion by test3211234** [lapce/lapce#3363] [Q&A]
I'm curious. I don't know how the commands are used/actually called, but I see that they're enums. I don't see how the when-clauses (in the .toml keymap files) are checked. I don't see how the proper functions are called when the right key is pressed. I see so many different run_command functions and it confuses me a lot. I would appreciate if a contributor took a bit of time to explain how all of that works to me.

I'm wondering all this because I was thinking of creating a VSCode-like keybinding crate and Lapce has similar things. But it's too confusing for me to take inspiration from. Do you think this is a good idea? Thanks!

**Comment by panekj** [comment]
> I'm wondering all this because I was thinking of creating a VSCode-like keybinding crate and Lapce has similar things. But it's too confusing for me to take inspiration from. Do you think this is a good idea?

For what would that crate be? If it's for Lapce, then that's not possible since Lapce does not handle additional functionality via crates.

**Reply by test3211234** [reply]
I was just wondering in general if it's a good idea for a crate to easily call the right functions when a key is pressed, with a context struct passed in with a when clause context like VS-code. I can see that this kind of thing might be too program-specific though.
